google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,474 @@
+          if (outstanding_write) {
+            outstanding_write_future = in_progress_write_->get_future();
+          }
+
+          // moving will change `stream_` to `nullptr`
+          stream = std::move(stream_);
+        }
+
+        if (outstanding_write) {
+          outstanding_write_future.then(
+              [this, stream = std::move(stream)](future<void>) {
+                FinishOnStreamFail(std::move(stream));
+              });
+        } else {
+          FinishOnStreamFail(std::move(stream));
+        }
+      }
+
+      return read_reinit_done_->get_future().then([](future<void>) {
+        return make_ready_future(absl::optional<ResponseType>());
+      });
+    });
+  }
+
+  future<bool> Write(Request const& r, grpc::WriteOptions o) override {
+    future<bool> write_future;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (stream_state_ == State::kShutdown) {
+        return make_ready_future(false);
+      }
+      if (stream_state_ == State::kRetrying) {
+        write_reinit_done_.emplace(absl::make_optional(promise<bool>()));
+        return write_reinit_done_->get_future();
+      }
+      write_future = stream_->Write(r, o);
+      in_progress_write_.emplace(absl::make_optional(promise<void>()));
+    }
+
+    return write_future.then([this](future<bool> write_future) {
+      {
+        std::lock_guard<std::mutex> g{mu_};
+        in_progress_write_->set_value();
+        in_progress_write_ = absl::optional<promise<void>>();
+      }
+
+      if (write_fu.get()) {
+        return make_ready_future(true);
+      }
+
+      bool retrying = false;
+
+      {
+        std::lock_guard<std::mutex> g{mu_};
+        write_reinit_done_.emplace(absl::make_optional(promise<void>()));
+        retrying = stream_state_ == State::kRetrying;
+      }
+
+      if (!retrying) {
+        bool outstanding_read = false;
+        std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+            stream;
+        future<void> outstanding_read_future;
+        {
+          std::lock_guard<std::mutex> g{mu_};
+          stream_state_ = State::kRetrying;
+
+          outstanding_read = in_progress_read_.has_value();
+          if (outstanding_read) {
+            outstanding_read_future = in_progress_read_->get_future();
+          }
+
+          stream = std::move(stream_);
+        }
+
+        if (outstanding_read) {
+          outstanding_read_future.then(
+              [this, stream = std::move(stream)](future<void>) {
+                FinishOnStreamFail(std::move(stream));
+              });
+        } else {
+          FinishOnStreamFail(std::move(stream));
+        }
+      }
+
+      return write_reinit_done_->get_future().then(
+          [](future<void>) { return make_ready_future(bool); });
+    });
+  }
+
+  future<Status> Finish() override {
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (stream_state_ == State::kShutdown) {
```

Comment:
you still need to handle the case if the stream is currently retrying, but is outstanding to the user reinitialization handler I think?

Reply:
I don't think so? the user can only call `Finish` if there are no pending `Read`s or `Write`s on their end. If there are no pending futures on the user's end, then we can't be in a retry loop internally. At least that's what I thought the implementation should look like/is looking like.